wpilibc/src/main/native/include/frc/shuffleboard/ShuffleboardContainer.h

```diff
@@ -0,0 +1,311 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <networktables/NetworkTableEntry.h>
+#include <networktables/NetworkTableValue.h>
+#include <wpi/ArrayRef.h>
+#include <wpi/SmallSet.h>
+#include <wpi/StringMap.h>
+#include <wpi/StringRef.h>
+
+#include "frc/shuffleboard/ShuffleboardComponentBase.h"
+#include "frc/shuffleboard/ShuffleboardValue.h"
+
+namespace frc {
+
+class ComplexWidget;
+class Sendable;
+class ShuffleboardLayout;
+class SimpleWidget;
+
+/**
+ * Common interface for objects that can contain shuffleboard components.
+ */
+class ShuffleboardContainer : public virtual ShuffleboardValue {
```

Comment:
Is virtual inheritance necessary?  We currently don't have it anywhere else in wpilib.

Reply:
It's used here because ShuffleboardComponentBase and ShuffleboardContainer both inherit from this class and ShuffleboardLayout inherits both of them. I made ShuffleboardValue a data class to avoid duplication of that data member and make it accessible from the base class since I needed a base class for that purpose anyway.

Reply:
Okay.  I guess we just don't typically have this complex of class hierarchies.